File: app/src/main/kotlin/br/com/victorwads/goldenunicorn/data/models/DebitRegistry.kt
package br.com.victorwads.goldenunicorn.data.models

import com.google.firebase.firestore.DocumentId
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.PropertyName
import java.util.Date

data class DebitRegistry(
    @DocumentId val id: String? = null,
    @PropertyName("bankId") val bankId: String,

    @PropertyName("name") val name: String,
    @PropertyName("type") val type: Type = Type.DEFAULT,
    @PropertyName("date") val date: Date,
    @PropertyName("done") val done: Boolean = false,
    @PropertyName("metadata") val metadata: Metadata? = null,

    @PropertyName("createdAt")
    val createdAt: FieldValue? = FieldValue.serverTimestamp(),
) {
    @PropertyName("updatedAt")
    val updatedAt: FieldValue = FieldValue.serverTimestamp()

    enum class Type {
        DEFAULT, FIXED
    }

    data class Metadata(
        @PropertyName("fixedBillId")
        val installmentBillId: String? = null,
        @PropertyName("recurrentBillId")
        val recurrentBillId: String? = null,
    )
}

File: iOS/Sources/Data/models/DebitRegistry.swift

import Foundation
import FirebaseFirestore

struct DebitRegistry: Codable, Identifiable {
    @DocumentID var id: String? = nil
    var bankId: String
    var categoryId: String

    var name: String = ""
    var amount: Double = 0
    var date: Date = Date()
    var type: Type = .default
    var done: Bool = false

    var metadata: Metadata = Metadata()
    var createdAt: Date = Date()
    var updatedAt: Date = Date()
    
    enum CodingKeys: String, CodingKey {
        case id
        case bankId = "bankId"
        case categoryId = "categoryId"
        
        case name = "name"
        case amount = "amount"
        case date = "date"
        case type = "type"
        case done = "done"
        
        case metadata = "metadata"
        case createdAt = "createdAt"
        case updatedAt = "updatedAt"
    }
    
    enum `Type`: String, Codable {
        case `default` = "DEFAULT"
        case fixed = "FIXED"
    }
    
    struct Metadata: Codable {
        var fixedBillId: String? = nil
        
        enum CodingKeys: String, CodingKey {
            case fixedBillId = "fixedBillId"
        }
    }
}
